src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComMethodInfo.cs

```diff
@@ -0,0 +1,120 @@
+            }
+
+            private static Diagnostic? GetDiagnosticIfInvalidMethodForGeneration(MethodDeclarationSyntax comMethodDeclaringSyntax, IMethodSymbol method)
+            {
+                // Verify the method has no generic types or defined implementation
+                // and is not marked static or sealed
+                if (comMethodDeclaringSyntax.TypeParameterList is not null
+                    || comMethodDeclaringSyntax.Body is not null
+                    || comMethodDeclaringSyntax.Modifiers.Any(SyntaxKind.SealedKeyword))
+                {
+                    return Diagnostic.Create(GeneratorDiagnostics.InvalidAttributedMethodSignature, comMethodDeclaringSyntax.Identifier.GetLocation(), method.Name);
+                }
+
+                // Verify the method does not have a ref return
+                if (method.ReturnsByRef || method.ReturnsByRefReadonly)
+                {
+                    return Diagnostic.Create(GeneratorDiagnostics.ReturnConfigurationNotSupported, comMethodDeclaringSyntax.Identifier.GetLocation(), "ref return", method.ToDisplayString());
+                }
+
+                return null;
+            }
+
+            private static bool IsComMethodCandidate(ISymbol member)
+            {
+                return member.Kind == SymbolKind.Method && !member.IsStatic;
+            }
+
+            private static (ComMethodInfo?, IMethodSymbol, Diagnostic?) CalculateMethodInfo(ComInterfaceInfo ifaceContext, IMethodSymbol method, CancellationToken ct)
+            {
+                ct.ThrowIfCancellationRequested();
+                Debug.Assert(IsComMethodCandidate(method));
+
+                // We only support methods that are defined in the same partial interface definition as the
+                // [GeneratedComInterface] attribute.
+                // This restriction not only makes finding the syntax for a given method cheaper,
+                // but it also enables us to ensure that we can determine vtable method order easily.
+                Location interfaceLocation = ifaceContext.Declaration.GetLocation();
+                Location? methodLocationInAttributedInterfaceDeclaration = null;
+                foreach (var methodLocation in method.Locations)
+                {
+                    if (methodLocation.SourceTree == interfaceLocation.SourceTree
+                        && interfaceLocation.SourceSpan.Contains(methodLocation.SourceSpan))
+                    {
+                        methodLocationInAttributedInterfaceDeclaration = methodLocation;
+                        break;
+                    }
+                }
+                // TODO: this should cause a diagnostic
+                if (methodLocationInAttributedInterfaceDeclaration is null)
+                {
+                    throw new NotImplementedException($"Could not find location for method {method.ToDisplayString()} within the attributed declaration");
```

Comment:
Let's fix up these `NotImplementedException`s into diagnostics, and then I think we're good to go!

Reply:
With these being issues with our analysis and not necessarily user code, should the diagnostic be something like "We can't figure this out, file a bug if you think we should be able to"? And if so, should we do a new diagnostic ID, or ConfigurationNotSupported?